algorithms/linfa-logistic/src/lib.rs

```diff
@@ -93,79 +115,60 @@ impl<F: Float> LogisticRegression<F> {
 
     /// Set the normalization parameter `alpha` used for L2 normalization,
     /// defaults to `1.0`.
-    pub fn alpha(mut self, alpha: F) -> LogisticRegression<F> {
+    pub fn alpha(mut self, alpha: F) -> Self {
         self.alpha = alpha;
         self
     }
 
     /// Configure if an intercept should be fitted, defaults to `true`.
-    pub fn with_intercept(mut self, fit_intercept: bool) -> LogisticRegression<F> {
+    pub fn with_intercept(mut self, fit_intercept: bool) -> Self {
         self.fit_intercept = fit_intercept;
         self
     }
 
     /// Configure the maximum number of iterations that the solver should perform,
     /// defaults to `100`.
-    pub fn max_iterations(mut self, max_iterations: u64) -> LogisticRegression<F> {
+    pub fn max_iterations(mut self, max_iterations: u64) -> Self {
         self.max_iterations = max_iterations;
         self
     }
 
     /// Configure the minimum change to the gradient to continue the solver,
     /// defaults to `1e-4`.
-    pub fn gradient_tolerance(mut self, gradient_tolerance: F) -> LogisticRegression<F> {
+    pub fn gradient_tolerance(mut self, gradient_tolerance: F) -> Self {
         self.gradient_tolerance = gradient_tolerance;
         self
     }
 
     /// Configure the initial parameters from where the optimization starts.
-    /// The `params` array must have the same size as the number of columns of
-    /// the feature matrix `x` passed to the `fit` method
-    pub fn initial_params(mut self, params: Array1<F>, intercept: F) -> LogisticRegression<F> {
-        self.initial_params = Some((params, intercept));
+    /// The `params` array must have at least the same number of rows as there are columns on the
+    /// feature matrix `x` passed to the `fit` method
+    pub fn initial_params(mut self, params: Array<F, D>) -> Self {
```

Comment:
doesn't this try to achieve what we are doing with `FitWith`? (i.e. you can also initialize a linear model with these parameters and call `fit_with` on that)

Reply:
yes, but I think it's more intuitive if a fitted model can be constructed (i.e. having a `FittedLogisticRegression::new` function) with a given intercept and hyperplane. You can then update those initial weights with `FitWith` instead of putting them to the hyperparameters.

Reply:
Both `FittedLogisticRegression` and `MultiFittedLogisticRegression` require the set of labels to be specified at construction, which isn't info that's available before fitting. If we make `FitWith` perform a single iteration then the users will have to write their own loops and stopping conditions, which is more work. We could make `FitWith` perform the complete algorithm to solve this. Also, if we have the `Fit` version of the algorithm always start with param arrays of 0s, it'd be pretty much useless since the optimal starting params are random values, not 0s. I just don't think it makes sense to delegate the offline and online versions of the algorithm to the same impl. What we can do is have separate param types for the online and offline versions of the algorithm, both implementing `FitWith`. The offline impl would run the entire algorithm while the online impl runs a single step. This makes sense because the offline algorithm requires a few hyperparams that the online version doesn't need.